asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go

```diff
@@ -0,0 +1,589 @@
+		return fmt.Errorf("Failed to get asset: " + err.Error())
+	} else if exists {
+		return fmt.Errorf("This asset already exists: " + ID)
+	}
+
+	// ==== Create asset object and marshal to JSON ====
+	objectType := "asset"
+	asset := &asset{objectType, ID, color, size, owner, appraisedValue}
+	assetJSONasBytes, err := json.Marshal(asset)
+	if err != nil {
+		return err
+	}
+	//Alternatively, build the asset json string manually if you don't want to use struct marshalling
+	//assetJSONasString := `{"docType":"asset",  "name": "` + ID + `", "color": "` + color + `", "size": ` + strconv.Itoa(size) + `, "owner": "` + owner + `"}`
+	//assetJSONasBytes := []byte(str)
+
+	// === Save asset to state ===
+	err = ctx.GetStub().PutState(ID, assetJSONasBytes)
+	if err != nil {
+		return err
+	}
+
+	//  ==== Index the asset to enable color-based range queries, e.g. return all blue assets ====
+	//  An 'index' is a normal key/value entry in state.
+	//  The key is a composite key, with the elements that you want to range query on listed first.
+	//  In our case, the composite key is based on indexName~color~name.
+	//  This will enable very efficient state range queries based on composite keys matching indexName~color~*
+	indexName := "color~name"
+	colorNameIndexKey, err := ctx.GetStub().CreateCompositeKey(indexName, []string{asset.Color, asset.ID})
+	if err != nil {
+		return err
+	}
+	//  Save index entry to state. Only the key name is needed, no need to store a duplicate copy of the asset.
+	//  Note - passing a 'nil' value will effectively delete the key from state, therefore we pass null character as value
+	value := []byte{0x00}
+	ctx.GetStub().PutState(colorNameIndexKey, value)
+
+	// ==== Asset saved and indexed. Return success ====
+	return nil
+}
+
+// ===============================================
+// ReadAsset - read a asset from chaincode state
+// ===============================================
+func (t *SimpleChaincode) ReadAsset(ctx contractapi.TransactionContextInterface, name string) (*asset, error) {
+	valAsbytes, err := ctx.GetStub().GetState(name) //get the asset from chaincode state
+	if err != nil {
+		return nil, err
+	} else if valAsbytes == nil {
+		return nil, fmt.Errorf("%s does not exist", name)
+	}
+
+	asset := new(asset)
+	err = json.Unmarshal(valAsbytes, asset)
+	if err != nil {
+		return nil, err
+	}
+
+	return asset, nil
+}
+
+// ==================================================
+// DeleteAsset - remove a asset key/value pair from state
+// ==================================================
+func (t *SimpleChaincode) DeleteAsset(ctx contractapi.TransactionContextInterface, assetName string) error {
+	var jsonResp string
+	var assetJSON asset
+
+	// to maintain the color~name index, we need to read the asset first and get its color
+	valAsbytes, err := ctx.GetStub().GetState(assetName) //get the asset from chaincode state
+	if err != nil {
+		jsonResp = "{\"Error\":\"Failed to get state for " + assetName + "\"}"
```

Comment:
not sure what's the purpose to use jsonResp for error formatting. Can probably just use `fmt.Errorf`.

Reply:
🛠️  this is from when a pb.Response was being returned.